ILEditing/MechanicILChanges.cs

```diff
@@ -772,332 +774,497 @@ private static void DrawForegroundParticles(Terraria.On_Main.orig_DrawInfernoRin
             {
-                case 1:
-                    Main.tileBatch.Draw(slopeTexture, position, liquidSize, colors, Vector2.Zero, 1f, SpriteEffects.None);
-                    break;
-                case 2:
-                    Main.tileBatch.Draw(slopeTexture, position, liquidSize, colors, Vector2.Zero, 1f, SpriteEffects.None);
-                    break;
-                case 3:
-                    Main.tileBatch.Draw(slopeTexture, position, liquidSize, colors, Vector2.Zero, 1f, SpriteEffects.None);
-                    break;
-                case 4:
-                    Main.tileBatch.Draw(slopeTexture, position, liquidSize, colors, Vector2.Zero, 1f, SpriteEffects.None);
-                    break;
+                LogFailure("Draw lavas to captures", "Could not locate the second liquid capture drawing");
+                return;
+            }
+            cursor.EmitDelegate(() => {
+                LavaRendering.instance.DrawLiquid(bg: false, CalamityMod.LavaStyle);
+            });
+            if (!cursor.TryGotoNext(MoveType.After, i => i.MatchLdloc2(), i => i.MatchStsfld<Main>("liquidAlpha")))
+            {
+                LogFailure("Draw lavas to captures", "Could not locate water style value returner");
+                return;
             }
+            cursor.EmitDelegate(() => {
+                CalamityMod.lavaAlpha = LavaRendering.alphaSave;
+            });
         }
 
-        private static void ChangeWaterQuadColors(ILContext il)
+        private void AddTileLiquidDrawing(ILContext il)
         {
             ILCursor cursor = new ILCursor(il);
-
-            if (!cursor.TryGotoNext(c => c.MatchLdfld<LiquidRenderer>("_liquidTextures")))
+            if (!cursor.TryGotoNext(MoveType.After, i => i.MatchLdarg0(), i => i.MatchLdarg1(), i => i.MatchLdcI4(0), i => i.MatchLdarg(out _), i => i.MatchLdloc1(), i => i.MatchLdloc2(), i => i.MatchLdloc(12), i => i.MatchLdloc(13), i => i.MatchLdloc(14), i => i.MatchCall<TileDrawing>("DrawTile_LiquidBehindTile")))
             {
-                LogFailure("Custom Lava Drawing", "Could not locate the liquid texture array load.");
+                LogFailure("Tile Lava Drawing", "Could not locate the drawing of Liquid Behind Tile drawing");
                 return;
             }
+            cursor.EmitLdloc1();
+            cursor.EmitLdloc2();
+            cursor.EmitLdloc(12);
+            cursor.EmitLdloc(13);
+            cursor.EmitLdloc(14);
+            cursor.EmitDelegate((Microsoft.Xna.Framework.Vector2 unscaledPosition, Microsoft.Xna.Framework.Vector2 vector, int j, int i, Terraria.Tile tile) => {
+                LavaRendering.instance.DrawTile_LiquidBehindTile(solidLayer: false, inFrontOfPlayers: false, -1, unscaledPosition, vector, j, i, tile);
+            });
```

Comment:
Historically, we don't like using hardcoded local indices because they've been brittle across Terraria updates. However, given 1.4.5 is going to be apocalyptic regardless of what we do, it's not crucial.